0000-py-version-support.md

```diff
@@ -0,0 +1,244 @@
+- Feature Name: Python Version Support
+- Start Date: 2020-03-24
+- SEP Status: Draft
+- SEP PR: (leave this empty)
+- Salt Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This SEP is being created to add a policy around Python versions Salt will support and
+detail how we will support Python versions going forward. This policy change is to align Salt Releases
+with Python's End of Life schedule.
+
+# Motivation
+[motivation]: #motivation
+
+Documentation and Communication
+-------------------------------
+One of the motivating factors around this policy change is to make it more clear what our Python
+version support is for each Salt releases. If it is clearly defined, then packagers, community members,
+and Salt users will know when a Python version will be dropped for a given Salt release.
+In the past we had never clearly defined how our Python version support lined up with Python's
+End of Life schedule. We previously made a best effort to support the Python versions that lined
+up with the operating systems Python versions, but never clearly defined this and did not always
+follow this in every case. For example, when we dropped Python 2.6 support we still supported
+Redhat 6, which had Python 2.6 installed by default. To continue supporting Redhat 6 while also
+removing support for the End of Life Python 2.6 version, we included Python 2.7 in our packages
+for that operating system.
+
+Including Python in Packages
+----------------------------
+Another motivating factor to make this policy change is our shift to including Python in our
+packages. For the Sodium release we will be using salt-bin (https://github.com/saltstack/salt-bin) to
+package for our Fully Supported Platforms.
+(http://get.saltstack.com/rs/304-PHQ-615/images/SaltStack-Supported-Operating-Systems.pdf)
+This will allow us to easily continue supporting operating systems that include End of Life Python
+versions by default by providing the Python binary in the Salt package included in a Salt release.
```

Comment:
What about when the OS doesn't have an old python? I want to use that, not bloat salt with its own version.
If the distribution doesn't provide its own newer python repo, it would be better to at least package python separately (or use official python repos) than to include it in the salt package.

Reply:
I didn't want to focus on the details around the shift/change to using salt-bin for Salt as there will be a separate SEP for this and this SEP focuses on what versions of python we will support. But yes we will be shifting to including python as a part of the package and don't intend on separating it out. There will be more details in that SEP and I'm not currently assigned to working on that team that is doing the research and work around it currently, so that SEP will have more details than i can provide from the team working on it. I would suggest building your own packages or using pip if this is the approach you want to take for one of the operating systems we package for.

Reply:
> don't intend on separating it out

If you don't intend to change the plan, why go through the SEP process?

Reply:
Apologies, didn't mean to make it sound like I wasn't willing to adjust, rather that this conversation should take place in the other SEP. That SEP will include more information around the actual implementation. The individuals working on that project are still working on research and implementation details they need to work out before a SEP with thorough details can be written. The important thing to note is a supported python version will be included to install, but the details around implementation will be included in a future SEP.

Reply:
@thatch45, @cro or @dmurphy18  can  elaborate here though as they are more aware of the implementation details here. My understanding is the plan is to include python in the package, but not sure if there is a way to not include it if a user so chooses.

Reply:
The official packages that SaltStack will ship will include Python, this solves so many issues that go well beyond just targetting the specific python version, including making sure that deps don't break under the release.
The official packages distributed by SaltStack will never be the ONLY way to install Salt, we will still push Salt to pypi and we will still make all sources and build tools available following open source guidelines. If you want to build and distribute system packages on your own, that follow different guidelines, we will strongly encourage it.
But the burden of packaging Salt for all distributions and situations has grown into such a significant burden that it slows testing, slows development, damages releases and introduces more bugs. The benefits we are gaining by following this process are amazingly clear.

Again, this change does not stop anyone from packaging Salt in a classical manner, but it does allow us to make the install and use of Salt much easier, for users, maintainers, distributors and more.

Finally, to answer the level of added bloat, the python distribution shipping with pop-build binaries, build using the configs in the salt-bin repo, is tiny. The python component ranges from just 7-15 megs depending on the platform. I am unaware of any remotely modern computing system, that runs a salt minion, IOT devices included, where this is not a negligible hit to the total storage space on the device.

Reply:
The plan is to include Python 3.7 as a minimum in salt-bin, since this reduces the number of dependencies that have to be managed and also deal with a known quantity. Python 3.7 is picked due to PEP 538 and 540 which issues with locales on some platforms and containers only providing for locale 'C' and POSIX.  Some of those platforms have Python 3.6 and PEP 538 is backported but ran into issues with traceback stackframe being output and other logging issues with exceptions resulting in OOM Killer's being called, which were completely resolved using Python 3.7.  As platforms have new versions of Python, for example: Ubuntu 20.04 and Python 3.8, then salt-bin will leverage Python 3.8.  Salt-bin leverages PyInstaller and picking a self-contained Python version allows for consistency across revisions to a platform.  One only needs to review the changes to Redhat 7 last summer with the changes between Python 3.4 to Python 3.6 and the ensuing upsets in EPEL and rpm naming, to isolate from those issues with Redhat or other platforms.